docs/Dialects/Verif.md

```diff
@@ -4,6 +4,322 @@ This dialect provides a collection of operations to express various verification
+
+These contracts can then be used in two key ways:
+
+1.  A contract can be _verified_ to hold by turning `require`s into `assume`s and `ensure`s into `assert`s.
+    Doing so verifies that a circuit upholds the contract by placing asserts on the values it produces, and by placing assumes on the input values the circuit sees.
+    In a nutshell, this checks that, assuming the inputs honor the contract, the circuit also upholds the contract.
+    This check can be done very efficiently by creating `verif.formal` ops to verify each contract.
+
+2.  A contract can be _assumed_ to hold by turning `require`s into `assert`s and `ensure`s into `assume`s.
+    Doing so verifies that the inputs fed into a circuit uphold the contract, such that the outputs can be assumed to have the promised values.
+    In a nutshell, this checks that the inputs to a circuit honor the contract and therefore the circuit can be assumed to uphold the contract.
+    Assuming a contract can often eliminate large parts of the circuit's actual implementation, since the contracts tend to be a simpler description of a circuit's functionality.
+
+### Multiply-by-9 Example
+
+Consider the following example of a HW module that computes `9 * a` using a left-shift and an addition.
+Note that this is using `verif.ensure_equal` as a standin for `verif.ensure(comb.icmp eq)`.
+The module's output `%z` is the result of a `verif.contract` operation.
+Contracts can be completely ignored by simply passing through their operands, `%1` in this case, to their results.
+This is their normal interpretation outside a formal verification flow, for example for synthesis.
+
+```mlir
+hw.module @Mul9(in %a: i42, out z: i42) {
+  // Compute 9*a as (a<<3)+a.
+  %c3_i42 = hw.constant 3 : i42
+  %0 = comb.shl %a, %c3_i42 : i42
+  %1 = comb.add %a, %0 : i42
+
+  // Contract to check that the circuit actually produces 9*a.
+  %z = verif.contract %1 : i42 {
+    %c9_i42 = hw.constant 9 : i42
+    %a9 = comb.mul %a, %c9_i42 : i42
+    verif.ensure_equal %z, %a9
+  }
+
+  hw.output %z : i42
+}
+```
+
+To prove that the contract holds, it can be pulled out into a `verif.formal` op along with all ops in its fan-in cone.
+The contract's results are then replaced with its operands, the body is inlined, and `ensure` ops are replaced with `assert` and `require` ops are replaced with `assume`.
+Running this formal test will prove that the `(a<<3)+a` implemented by the module is indeed the same as the `9*a` promised by the contract.
+
+```mlir
+verif.formal @Mul9_ProveContract {
+  %a = verif.symbolic_value : i42
+
+  %c3_i42 = hw.constant 3 : i42
+  %0 = comb.shl %a, %c3_i42 : i42
+  %1 = comb.add %a, %0 : i42
+
+  // Contract inlined with ensure -> assert, %z -> %1
+  %c9_i42 = hw.constant 9 : i42
+  %a9 = comb.mul %a, %c9_i42 : i42
+  verif.assert_equal %1, %a9 : i42
+}
+```
+
+Once the contract is proven, it can be assumed to hold everywhere.
```

Comment:
I honestly prefer the term "checked" rather than proven, because we aren't actually proving anything, we're just doing a heuristic-based search of the input space for a satisfiability problem. Proving would require something like an inductive proof, which we don't actually do ^^"

Reply:
We could though :smirk: But yeah, you're right.